### PR TITLE
Make Session.chdir() a context manager

### LINKED ISSUE #512: Run commands from within a specific directory
**How would this feature be useful?**

Since many commands have specific usages of the current working directory, 
running commands from within a specific directory would smooth integrating nox
is existing projects.

**Describe the solution you'd like**

```python
from pathlib import Path

import nox


@nox.session()
def test(session: nox.Session):
    session.install('pytest')
    python_package_root = Path("somewhere/deep/in/monorepo")
    session.run(sys.executable, "-m", "pytest", cwd=python_package_root)
```

**Describe alternatives you've considered**

```python
import os
from pathlib import Path

import nox


@nox.session()
def test(session: nox.Session):
    session.install('pytest')
    python_package_root = Path("somewhere/deep/in/monorepo")
    before_cwd = os.getcwd()
    os.chdir(python_package_root)
    session.run(sys.executable, "-m", "pytest")
    os.chdir(before_cwd)
```

### PR DESCRIPTION
**Pull request by Tolker-KU**
Closes #512. Also see the discussion here: #513.

I'm a bit unsure if the changes to the docs are sufficient. Please suggest changes as you see fit.

### CODE REVIEW

File: nox/sessions.py [hunk 1]
```diff
@@ -210,10 +212,37 @@ def invoked_from(self) -> str:
         """
         return self._runner.global_config.invoked_from
 
-    def chdir(self, dir: Union[str, os.PathLike]) -> None:
-        """Change the current working directory."""
+    class _WorkingDirContext:
```
Comment (henryiii) [inline nox/sessions.py:215]: I really dislike nested classes and functions. The main reason to do it is for capture, which is a bad, unreadable reason. Would this be better as an un-nested class?

Comment (Tolker-KU) [reply nox/sessions.py]: Certainly not. It is one of my bad habits. I've moved it out in the global scope 👍

**Comment by FollowTheProcess** [comment]
Could this be simplified further to a functional context manager? i.e. you wouldn't need the `WorkingDirContext` class at all?

Something like:

```python
@contextlib.contextmanager
def chdir(self, to: str) -> Iterator:
    original = os.getcwd()
    try:
        os.chdir(to)
        yield
    finally:
        os.chdir(original)
```

I haven't done much with context managers tbh so I'm not 100% that these are equivalent

**Comment by Tolker-KU** [comment]
> Could this be simplified further to a functional context manager? i.e. you wouldn't need the `WorkingDirContext` class at all?
> 
> Something like:
> 
> ```python
> @contextlib.contextmanager
> def chdir(self, to: str) -> Iterator:
>     original = os.getcwd()
>     try:
>         os.chdir(to)
>         yield
>     finally:
>         os.chdir(original)
> ```
> 
> I haven't done much with context managers tbh so I'm not 100% that these are equivalent

As far as I know the problem with this is that it would break the current functionality. 

```python
@nox.session()
def some_session(session):
    session.chdir("path/to/somewhere")
    # Clean has been run here, so we are back in old working dir
```

Using the class approach we are able to keep the current functionality at the cost of some complexity.

**Review by cjolowicz — APPROVED** [review]

**Comment by FollowTheProcess** [comment]
> > Could this be simplified further to a functional context manager? i.e. you wouldn't need the `WorkingDirContext` class at all?
> 
> > 
> 
> > Something like:
> 
> > 
> 
> > ```python
> 
> > @contextlib.contextmanager
> 
> > def chdir(self, to: str) -> Iterator:
> 
> >     original = os.getcwd()
> 
> >     try:
> 
> >         os.chdir(to)
> 
> >         yield
> 
> >     finally:
> 
> >         os.chdir(original)
> 
> > ```
> 
> > 
> 
> > I haven't done much with context managers tbh so I'm not 100% that these are equivalent
> 
> 
> 
> As far as I know the problem with this is that it would break the current functionality. 
> 
> 
> 
> ```python
> 
> @nox.session()
> 
> def some_session(session):
> 
>     session.chdir("path/to/somewhere")
> 
>     # Clean has been run here, so we are back in old working dir
> 
> ```
> 
> 
> 
> Using the class approach we are able to keep the current functionality at the cost of some complexity.

Cool, makes sense, thanks! This looks good to me then

**Review by FollowTheProcess — APPROVED** [review]

**Pull request merged**